syntax = "proto3";


service StatsService {
    rpc GetPostStats (PostStatsRequest) returns (PostStatsOrErrorReply) {}

    rpc GetTopPosts (TopPostsRequest) returns (TopPostsStatsOrErrorReply) {}

    rpc GetTopUsers (EmptyRequest) returns (TopUsersOrErrorReply) {}
}

message EmptyRequest {
}

message PostStatsRequest {
    int32 postId = 1;
}

message TopPostsRequest {
    bool isLikes = 1;
}

message OnePostStat {
    int32 id = 1;
    int32 stat = 2;
}

message PostStats {
    int32 id = 1;
    int32 likes = 3;
    int32 views = 4;
}

message UserStats {
    string login = 2;
    int32 likes = 3;
}

message PostStatsOrErrorReply {
    bool isOk = 1;
    optional PostStats postStats = 2;
    optional int32 errorCode = 3;
    optional string errorText = 4;
}

message TopPostsStatsOrErrorReply {
    bool isOk = 1;
    repeated OnePostStat posts = 2;
    optional int32 errorCode = 3;
    optional string errorText = 4;
}

message TopUsersOrErrorReply {
    bool isOk = 1;
    repeated UserStats users = 2;
    optional int32 errorCode = 3;
    optional string errorText = 4;
}
